acdc/TLACDCExperiment.py

```diff
@@ -384,32 +384,39 @@ def add_all_sender_hooks(self, reset=True, cache="first", skip_direct_computatio
     def setup_second_cache(self):
         if self.verbose:
             print("Adding sender hooks...")
-        
+
         self.model.reset_hooks()
-        self.model.cache_all(self.global_cache.second_cache)
 
-        if self.verbose:
-            print("Now corrupting things..")
+        if self.zero_ablation:
+            # to calculate the inputs to each model component, 
+            # we need zero out all the outputs into the residual stream
+
+            # all hooknames that output into the residual stream
+            hook_name_substrings = ["attn_result", "mlp_out", "hook_embed"]
+            if self.use_pos_embed:
+                hook_name_substrings.append("hook_pos_embed")
+
+            # add hooks to zero out all these hook points
+            hook_name_bool_function = lambda hook_name: any([hook_name_substring in hook_name for hook_name_substring in hook_name_substrings])
+            self.model.add_hook(
+                name = hook_name_bool_function,
+                hook = lambda z, hook: torch.zeros_like(z),
```

Comment:
Does this then zero out the biases from the attention heads?

Reply:
No, since `attn_result` does not include attention head biases as discussed

Reply:
Right, and I guess that's the correct thing to do with zero ablation.